### DISCUSSION TITLE
Helix mode for Obsidian (and other things) via CodeMirror

### DISCUSSION THREAD

**Discussion by DumbDurian** [helix-editor/helix#8386] [Ideas]
Mainly looking for some thoughts on the feasibility of building something like a 'helix-mode' for (mainly, but not exclusively) Obsidian.

Obsidian uses https://codemirror.net/, more specifically https://github.com/replit/codemirror-vim to offer a vim-mode within Obsidian.

I know neither of the tools' internals very well, so I am looking for the assessment of others! As far as I can tell most of the logic lives in https://github.com/replit/codemirror-vim/blob/master/src/vim.js

I am not sure if this is auto-generated, but either way, one would probably have to translate the helix keymap in the same way I assume. Is there an equivalent keymap file in the helix repository? https://github.com/helix-editor/helix/blob/master/helix-term/src/keymap.rs seems rather small, are there any other files I should have a look at?

I think having a CodeMirror like implementation for helix could go a long way for adoption as other things like for example typst.app are depending on CodeMirror as well.

Please provide some thoughts, feedback or ideas on how to possibly tackle this problem!

**Comment by Omnikar** [comment]
> Is there an equivalent keymap file in the helix repository? https://github.com/helix-editor/helix/blob/master/helix-term/src/keymap.rs seems rather small, are there any other files I should have a look at?

Sounds like you're looking for https://github.com/helix-editor/helix/blob/master/helix-term/src/keymap/default.rs

**Comment by zetashift** [comment]
A lot of the Electron/online editor programs that have a vim mode use CodeMirror in some way. So I believe some plugin for CodeMirror would be the best bet. 
One could also ask in the CodeMirror forums/repo what the best way to tackle this problem is, or even Obsidian forums. The maintainers of these projects are active.

**Comment by claudemuller** [comment]
@DumbDurian I use Jetbrains' IDEs at work and, having used Vim for many years, use a plugin called [IdeaVim](https://github.com/JetBrains/ideavim) there. Recently I have switched to Helix and have been trying to Helix-ify all the places that I used Vim.

In IdeaVim I have a non-complete frankenstein of other people's keymaps as well as my own [here](https://github.com/claudemuller/dot-files/blob/master/ideavim/.ideavimrc). I've just started trying to do the same with the [vimrc-support plugin](https://github.com/esm7/obsidian-vimrc-support), but a cleaner solution would be to create/use a Helix plugin that works similar to vimrc.

My Obsidian notes are private, but a couple of things I've gotten working this morning are `d -> delete` and `x -> select line`. This is my `.obsidian.vimrc` for these:
```bash
nnoremap d x
vnoremap d x
nnoremap x 0v$
vnoremap x j
```

Have you discovered anything else that would help in bringing Helix to Obsidian?

**Comment by LifeWorks** [comment]
Do you happen to have any updates on this?

**Comment by jrvidal** [comment]
FYI I've given this a try, and I have a very rough initial version at https://www.npmjs.com/package/codemirror-helix

Demo: [https://_rvidal.gitlab.io/codemirror-helix/](https://_rvidal.gitlab.io/codemirror-helix/)

**Reply by LifeWorks** [reply]
Is there any guidance about how to install this to obsidian?

**Reply by jrvidal** [reply]
Not really, I don't know anything about Obsidian itself, sorry.

**Reply by Sinono3** [reply]
I'm working on a simple plugin to build this into Obsidian. I'll push to a repo in an hour or two

**Reply by Sinono3** [reply]
Huge thanks for making this. Here's the plugin: https://github.com/Sinono3/obsidian-helix

**Reply by johannesrave** [reply]
so cool of both of you! thanks!

**Reply by shemraz** [reply]
thank you so much for this, you guys are amazing

**Reply by hollisticated-horse** [reply]
awesome, cheers @Sinono3 

Any news on the plugin coming out of beta ?

**Reply by Sinono3** [reply]
Hi. Actually, I'm looking for a person to maintain the project. https://github.com/Sinono3/obsidian-helix/issues/18

**Comment by LifeWorks** [comment]
> Huge thanks for making this. Here's the plugin: https://github.com/Sinono3/obsidian-helix

Thanks very much for making it easier. I got an error message while installing using BART. I created an issue for it in your repository. There is not problem with manual installation.

**Reply by Sinono3** [reply]
Fixed on [0.1.0](https://github.com/Sinono3/obsidian-helix/releases/tag/0.1.0)!

**Comment by noor-tg** [comment]
`mi(` not working with `obsidian-helix` plugin 
@jrvidal @Sinono3

**Comment by alper** [comment]
I would be happy here to have an easy way from Obsidian to:

* Pop a new terminal window
* Run Helix in the vault workin directory
* Open the current open note(s) in Helix as buffers

Would that be possible?

**Reply by johannesrave** [reply]
yes, i have this set up with the hot key `ctrl+shift+h` using the "Open With" obsidian-plugin.

you need to make it call your terminal and start helix in there:

<img width="1301" height="481" alt="image" src="https://github.com/user-attachments/assets/c5140c1c-ee0b-4a3c-a2d5-688408fef7ad" />


you can then configure the hotkey for it inside the general obsidian settings
<img width="1187" height="155" alt="image" src="https://github.com/user-attachments/assets/afa223cb-b8e2-4c3a-9526-7cfbaad9f55b" />

**Reply by alper** [reply]
I don't seem to be able to do the same with Ghostty.

**Reply by johannesrave** [reply]
i had a look and you need to configure `Open With` a bit weirdly, but it does work:
<img width="1466" height="104" alt="image" src="https://github.com/user-attachments/assets/dcce8a75-cff3-4a98-9870-131cd1efcfe3" />

so you use the path to your ghostty as the command, and the arguments are `-e` and `hx` and need to be comma separated - in effect `-e,hx`

**Reply by alper** [reply]
I think you're on Linux?

My Ghost says:
> On macOS, launching the terminal emulator from the CLI is not supported and only actions are supported.

**Reply by johannesrave** [reply]
i see - out of my depth then, sorry.

**Comment by littleblack111** [comment]
C-w not working with obsidian-helix plugin

**Reply by k4lizen** [reply]
then open an issue there?
